lib/integrations/awsoidc/listec2_test.go

```diff
@@ -0,0 +1,233 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awsoidc
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2Types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+type mockListEC2Client struct {
+	pageSize     int
+	accountID    string
+	ec2Instances []ec2Types.Instance
+}
+
+// GetCallerIdentity returns information about the caller identity.
+func (m *mockListEC2Client) GetCallerIdentity(ctx context.Context, params *sts.GetCallerIdentityInput, optFns ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error) {
+	return &sts.GetCallerIdentityOutput{
+		Account: &m.accountID,
+	}, nil
+}
+
+// Returns information about ec2 instances.
+// This API supports pagination.
+func (m mockListEC2Client) DescribeInstances(ctx context.Context, params *ec2.DescribeInstancesInput, optFns ...func(*ec2.Options)) (*ec2.DescribeInstancesOutput, error) {
+	requestedPage := 1
+
+	totalInstances := len(m.ec2Instances)
+
+	if params.NextToken != nil {
+		currentMarker, err := strconv.Atoi(*params.NextToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		requestedPage = currentMarker
+	}
+
+	sliceStart := m.pageSize * (requestedPage - 1)
+	sliceEnd := m.pageSize * requestedPage
+	if sliceEnd > totalInstances {
+		sliceEnd = totalInstances
+	}
+
+	ret := &ec2.DescribeInstancesOutput{
+		Reservations: []ec2Types.Reservation{{
+			Instances: m.ec2Instances[sliceStart:sliceEnd],
+		}},
+	}
+
+	if sliceEnd < totalInstances {
+		nextToken := fmt.Sprintf("%d", requestedPage+1)
+		ret.NextToken = stringPointer(nextToken)
+	}
+
+	return ret, nil
+}
+
+func TestListEC2(t *testing.T) {
+	ctx := context.Background()
+
+	noErrorFunc := func(err error) bool {
+		return err == nil
+	}
+
+	pageSize := 100
+	t.Run("pagination", func(t *testing.T) {
```

Comment:
This test focuses on validating lengths of the returned resource list. If we generated a different instance ID for each of those, we could also make an assertion about a selected one instance ID (say, the first one in the page) to make sure that the we actually return the right data. This might make this test more valuable.

Reply:
:+1:

Reply:
Done :pray: